### PR TITLE
[mdarea] Add mdarea package

### PR DESCRIPTION
**Pull request by toniopelo**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by toniopelo** [comment]
@jahudka You might be interested into this one?
I did it for myself so I figured I would publish it here but I have no real knowledge of your package.

**Comment by jahudka** [comment]
Hi @toniopelo, thanks so much for doing this! I'll look into it over the weekend. Just not sure what the best approach here is - should I just do a review, or should I make the changes that appear to be necessary? And in the latter case, how do I do that, seeing as this is your PR and your fork, so I probably don't have permissions to push here? Sorry, I'm not overly experienced with collaboration on large open-source projects 😅 thanks again!

**Comment by toniopelo** [comment]
@jahudka Hi, sorry for the delay! I think you can start by reviewing my code and mark the pieces of code you would like to update with you thoughts. You could even add suggestions with this button: 

image

This way I could directly commit your suggestions to my fork :)
What do you think ?

**Review by jahudka — CHANGES_REQUESTED** [review]
There's a lot of stuff I'd write differently - I prefer using types instead of interfaces for things that can't be implemented or extended (like the config structure), I'd probably `declare class MarkdownArea` with the full signature and export that rather than exporting an arbitrary interface whose only function is to construct _another_ interface.. but I think these are mostly a matter of preference and I'm not familiar with requirements for typings here on DefinitelyTyped, so I'm not gonna get into that ;-) if the typings work as they are it's good enough for me. Oh and I really dig the documentation included in the comments, so helpful not to have to leave my IDE to find out what an option means!

The only change that really needs to happen is the `export = ...` thing imo. Thanks for this again!

**Comment by toniopelo** [comment]
@jahudka I just applied your suggestion.
Feel free to make all the other suggestions that you would like to apply since you'll probably be the one editing this package latter on.
Also, you might want to add your name as Definition owner along mine this way: https://github.com/DefinitelyTyped/DefinitelyTyped#definition-owners

Have a nice day 😉

**Review by peterblazejewicz — COMMENTED** [review]
IMO, you will need to add support for global script usage:
https://github.com/jahudka/mdarea/blob/master/mdarea.js#L7
https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html#identifying-a-umd-library
as `MarkdownArea`
```ts
export as namespace MarkdownArea;
```

File: types/mdarea/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
+// Type definitions for mdarea 1.0
+// Project: https://github.com/jahudka/mdarea#readme
+// Definitions by: Antoine Laffargue <https://github.com/toniopelo>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
```
Comment (peterblazejewicz) [inline types/mdarea/index.d.ts:5]:
```suggestion
export as namespace MarkdownArea;

```

**Comment by elibarzilay** [comment]
@toniopelo: ping -- re @peterblazejewicz's comment, and is it all good re @jahudka's comments?  (Also, maybe add him as an owner too?)

**Review by peterblazejewicz — APPROVED** [review]

**Review by jahudka — APPROVED** [review]

**Comment by toniopelo** [comment]
@jahudka I still think that you should be listed as an author on this package, I you want to, please submit a suggestion and I will commit it :)
For the rest, LGTM 👍

File: types/mdarea/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
+// Type definitions for mdarea 1.0
+// Project: https://github.com/jahudka/mdarea#readme
+// Definitions by: Antoine Laffargue <https://github.com/toniopelo>
```
Comment (jahudka) [inline types/mdarea/index.d.ts:3]:
```suggestion
// Definitions by: Antoine Laffargue <https://github.com/toniopelo>, Dan Kadera <https://github.com/jahudka>
```

**Comment by jahudka** [comment]
sorry, I can be a real dummy sometimes - I know you've suggested I add myself a number of times before, just never got around to actually doing it..

File: types/mdarea/index.d.ts (hunk 2 above)
Comment (andrewbranch) [reply types/mdarea/index.d.ts:3]:
```suggestion
// Definitions by: Antoine Laffargue <https://github.com/toniopelo>
//                 Dan Kadera <https://github.com/jahudka>
```

**Review by andrewbranch — APPROVED** [review]

**Review by jahudka — APPROVED** [review]

**Review by peterblazejewicz — APPROVED** [review]

**Comment by toniopelo** [comment]
Ready to merge

**Pull request merged**